server/etcdserver/api/v3discovery/discovery.go

```diff
@@ -0,0 +1,544 @@
+	_, err := d.c.Put(ctx, memberKey, contents)
+	cancel()
+
+	if err != nil {
+		d.lg.Warn(
+			"failed to register members itself to the discovery service",
+			zap.String("discovery-url", d.durl),
+			zap.String("memberKey", memberKey),
+			zap.Error(err),
+		)
+		return d.registerSelfRetry(contents)
+	}
+
+	d.lg.Info(
+		"register member itself successfully",
+		zap.String("discovery-url", d.durl),
+		zap.String("memberKey", memberKey),
+		zap.String("memberInfo", contents),
+	)
+
+	return nil
+}
+
+func (d *discovery) waitNodes(members memberInfoMap, clusterSize int, rev int64) (memberInfoMap, error) {
+	// watch from the next revision
+	membersKeyPrefix := getMemberKeyPrefix(d.cluster)
+	w := d.c.Watch(context.Background(), membersKeyPrefix, clientv3.WithPrefix(), clientv3.WithRev(rev+1))
+
+	d.lg.Info(
+		"waiting for peers from discovery service",
+		zap.String("discovery-url", d.durl),
+		zap.Int("clusterSize", clusterSize),
+		zap.Int("found-peers", len(members)),
+	)
+
+	// waiting for peers until all needed peers are returned
+	for wresp := range w {
+		for _, ev := range wresp.Events {
+			mKey := strings.TrimSpace(string(ev.Kv.Key))
+			mValue := strings.TrimSpace(string(ev.Kv.Value))
+
+			if shouldIgnorePeer(members, mKey, mValue, d.cluster, d.lg, d.durl) {
+				continue
+			}
+
+			d.lg.Info(
+				"found peer from discovery service",
+				zap.String("discovery-url", d.durl),
+				zap.String("memberKey", mKey),
+				zap.String("memberInfo", mValue),
+			)
+
+			members[mKey] = memberInfo{peerURL: mValue, createRev: ev.Kv.CreateRevision}
+		}
+
+		if len(members) >= clusterSize {
+			break
+		}
+	}
+
+	d.lg.Info(
+		"found all needed peers from discovery service",
+		zap.String("discovery-url", d.durl),
+		zap.Int("clusterSize", clusterSize),
+		zap.Int("found-peers", len(members)),
+	)
+	return members, nil
+}
+
+func shouldIgnorePeer(members memberInfoMap, memberKey, memberInfo string, cluster string, lg *zap.Logger, durl string) bool {
+	membersKeyPrefix := getMemberKeyPrefix(cluster)
+
+	if !strings.HasPrefix(memberKey, membersKeyPrefix) {
+		// It should never happen.
+		lg.Panic(
+			"invalid peer registry key",
+			zap.String("discovery-url", durl),
+			zap.String("memberKey", memberKey),
+			zap.String("memberInfo", memberInfo),
+		)
+	}
+
+	if strings.IndexRune(memberInfo, '=') == -1 {
+		// It must be in the format "member1=http://127.0.0.1:2380".
+		lg.Warn(
+			"invalid peer info returned from discovery service",
+			zap.String("discovery-url", durl),
+			zap.String("memberKey", memberKey),
+			zap.String("memberInfo", memberInfo),
+		)
+		return true
+	}
+
+	if _, ok := members[memberKey]; ok {
+		lg.Warn(
+			"found duplicate peer from discovery service",
+			zap.String("discovery-url", durl),
+			zap.String("memberKey", memberKey),
+			zap.String("memberInfo", memberInfo),
+		)
+		return true
+	}
+
+	return false
+}
+
+func nodesToCluster(members memberInfoMap, clusterSize int, lg *zap.Logger) (string, error) {
```

Comment:
```suggestion
func (members memberInfoMap) Cluster(clusterSize int) (string, error) {
```
I would also recommend moving `clusterSize` verification to different place.

Reply:
The code (only the discovery.go) was refactored a little bit, please let me know if you still have comment on this.